docs/communicate-with-javascript.md

```diff
@@ -176,6 +176,162 @@ For a comprehensive understanding of abstract types and their usefulness, refer
+
+#### Pipe last
+
+Since version 4.01, OCaml includes a reverse application or "pipe" (`|>`)
+operator, an infix operator that applies the result from the previous expression
+the next function. As a backend for OCaml, Melange inherits this operator.
+
+The pipe operator could be implemented like this (the real implementation is a
+bit
+[different](https://github.com/ocaml/ocaml/blob/d9547617e8b14119beacafaa2546cbebfac1bfe5/stdlib/stdlib.ml#L48)):
+
+```ocaml
+let ( |> ) f g = g f
+```
+
+This operator is useful when multiple functions are applied to some value in
+sequence, with the output of each function becoming the input of the next (a
+_pipeline_).
+
+For example, assuming we have a function `square` defined as:
+
+```ocaml
+let square x = x * x
+```
+
+We are using it like:
+
+```ocaml
+let ten = succ (square 3)
+```
+
+The pipe operator allows to write the computation for `ten` in left-to-right
+order, as [it has left
+associativity](https://v2.ocaml.org/manual/expr.html#ss:precedence-and-associativity):
+
+```ocaml
+let ten = 3 |> square |> succ
+```
+
+When working with functions that can take multiple arguments, the pipe operator
+works best when the functions take the data we are processing as the last
+argument. For example:
+
+```ocaml
+let sum = List.fold_left ( + ) 0
+
+let sum_sq =
+  [ 1; 2; 3 ]
+  |> List.map square (* [1; 4; 9] *)
+  |> sum             (* 1 + 4 + 9 *)
+```
+
+The above example can be written concisely because the `List.map` function in
+the [OCaml standard library](https://v2.ocaml.org/api/Stdlib.List.html) takes
+the list as the second argument. This convention is sometimes referred to as
+"data-last", and is widely adopted in the OCaml ecosystem. Data-last and the
+pipe operator `|>` work great with currying, so they are a great fit for the
+language.
+
+However, there are some limitations when using data-last when it comes to error
+handling. In the given example, if we mistakenly used the wrong function:
+
+```ocaml
+let sum_sq =
+  [ 1; 2; 3 ]
+  |> List.map String.cat
+  |> sum
+```
+
+The compiler would rightfully raise an error:
+
+```text
+4 |   [ 1; 2; 3 ]
+        ^
+Error: This expression has type int but an expression was expected of type
+         string
+```
+
+Note that instead of telling us that we are passing the wrong function in
+`List.map` (`String.cat`), the error points to the list itself. This behavior
+aligns with the way type inference works, as the compiler infers types from left
+to right. Since `[ 1; 2; 3 ] |> List.map String.cat` is equivalent to `List.map
+String.cat [ 1; 2; 3 ]`, the type mismatch is detected when the list is type
+checked, after `String.cat` has been processed.
+
+With the goal of addressing this kind of limitations, Melange introduces the
+pipe first operator `|.`.
+
+#### Pipe first
+
+To overcome the constraints mentioned above and enable the generation of more
+performant JavaScript code, Melange introduces the pipe first operator `|.`.
```

Comment:
what's an example of a code snippet that is more performant with pipe first?

Reply:
https://github.com/rescript-lang/rescript-compiler/issues/1671

Reply:
I see. Updated docs in https://github.com/melange-re/melange-re.github.io/pull/50.